File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    temp = [i for i in range(n)]
    
    for i in range(n):
        for j in range(n):
            if i == j: continue
            if computers[i][j]:
                for k in range(n):
                    if temp[k] == temp[i]:
                        print(i, j, k)
                        temp[k] = temp[j]
    print(temp)
    return len(set(temp))

File: 프로그래머스/3/43162. 네트워크/네트워크.kt
class Solution {
    private fun find(x: Int, parents: MutableList<Int>): Int {
        if (x == parents[x]) return x
        return find(parents[x], parents)
    }
    
    private fun union(x: Int, y: Int, parents: MutableList<Int>) {
        val a = find(x, parents)
        val b = find(y, parents)
        
        if (a < b) {
            parents[b] = a
        } else {
            parents[a] = b
        }
    }
    
    fun solution(n: Int, computers: Array<IntArray>): Int {
        val parents = MutableList(size = n) { it }
        for (i in 0 until n) {
            for (j in 0 until n) {
                if (i != j && computers[i][j] == 1) {
                    union(i, j, parents)
                }
            }
        }
        val answers = mutableSetOf<Int>()
        for (i in 0 until n) {
            answers.add(find(i, parents))
        }
        
        return answers.size
    }
}